craft_parts/dirs.py

```diff
@@ -0,0 +1,51 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Definitions for project directories."""
+
+from pathlib import Path
+from typing import Union
+
+
+class ProjectDirs:
+    """The project main work directories.
+
+    :param work_dir: the parent directory containing the parts, prime and stage
```

Comment:
According to PEP257 it seems so, with the general class description in the class docstring and the constructor description under init (which many try to avoid because the short description for init is always a bit silly). I'll do as described in the PEP, minus list of methods and variables, to see how Sphinx handles it.

Reply:
Sphinx seems to prefer parameters documented in the class, contrary to what PEP257 recommends (but the PEP also recommends listing public methods and attributes in the class docstring which would result in redundancies in the processed docs). There are ways to make Sphinx concatenate both descriptions but the parameter list appears twice. Having the documentation formatted correctly is important, so unless @cjp256 or @facundobatista know a better way to handle this I'll leave it in the format that works for Sphinx.

Reply:
Generally speaking from everything I've read about it, it's best to pick one way or the other.  You can make it work either way, and probably both after wrestling with sphinx and autodoc, but idk.  The class to constructor is a 1:1 mapping so I don't care :D